content/blog/raxml_threading.md

```diff
@@ -0,0 +1,126 @@
+jobs on the node and letting the operating system decide the best use of its
+resources turns out effective both in terms of performance, and in the lack of
+tricky programming (or often, any programming at all) required to get that
+performance.
+
+Where there is threaded code, various R packages, or OpenMP, or Java threads
+all make it possible without too much fuss if you are careful. Or sometimes
+you might strike it lucky and find that the tool you want to use has a
+ready-made argument for threading - a command-line argument `-T 32` for
+instance, to make maximal use of all the cores on our speedy nodes. 
+Because why wouldn't you?
+
+# Why wouldn't you, indeed.
+
+A decade or so ago, our clusters consisted mainly of 8-core machines, which
+at the time felt like a tremendous breakthrough. The performance
+graphs we were used to seeing often looked something like this:
+
+{{< figure src="/img/raxml_08.png" alt="RAxML with up to 8 cores" >}}
+
+These are sample runs I made using our newest nodes this week with a
+bioinformatics tool called [RAxML](https://github.com/stamatak/standard-RAxML).
+It can be compiled to target SSE3, AVX, or AVX2 processor optimisations,
+and with threading enabled. It offers the convenient `-T threads` argument we
+mentioned earlier. 
+
+One of the threads (if you use more than one) acts as an
+administrator, which somewhat explains the lack of gain from 1 to 2 threads;
+after that, from 2 to 6 threads, we're around 90% efficient. Beyond that, the
+gains of more threads are diminishing. But a few years ago, that was where
+the graph ended. Now we have more cores, so let's throw them all at the
+problem...
+
+{{< figure src="/img/raxml_032.png" alt="RAxML with over 8 cores" >}}
+
+.. and we actually start to make things slower - using 32 cores performs
+comparably to using 8. There just isn't enough parallel work for all the
+threads to do; they spend more time waiting for the administrator to
+assign them work, than they spend executing that work.
+
+# Stacking
+
+So if throwing all a node's resources at a single job doesn't necessarily
+make that job faster (indeed, perhaps the opposite), then what if we try and
+maximise throughput instead? Let's try filling a 32-core node with as many 16, 8, or
+4-core jobs that will fit, and look for the best average time-per-job as
+we sdtack thme. For simplicity, I'll limit to AVX2.
+
+{{< figure src="/img/raxml_multi.png" alt="RAxML with jobs stacked on a node" >}}
+
+Here the blue bar shows the solo job we did earlier, where the job (however many
+threads) has the whole node to itself; the other bars show the jobs that we ran 
+simultaneously to fill the node up, to see how they are affecting by the stacking.
+
+The results are a bit confusing here and there; the 10-core is surprisingly 
+erratic, and needs some deeper investigation. The overhead of stacking up 4 and 8 core 
+jobs is a bit more than we might; perhaps those jobs are using more of the node than
```

Comment:
"than we might" => "than we might want"?

Reply:
Or "than we might expect"?

Reply:
thanks - "expect" was what I first had in mind - but have gone for "expect or want"